From 1e4367e7a2753031bb1583ea92b5f8b3bf389e6e Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Mon, 11 Sep 2023 15:28:08 +0200
Subject: [PATCH 0042/1695] drivers: regulator: GPIO controlled regulator

Implements a GPIO controlled regulators driver compliant with DT nodes
compatible with regulator-gpio. These regulators use GPIO pins to select
the voltage level. The implementation supports only dual voltage level
selection using a single pin. The DT bindings allows more pins to
select between more voltages but no known platform currently requires
that so we preferred the simplified case.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git abccde17cffc5f27b238c5f6070d0048dfc003ab drivers: regulator: GPIO controlled regulator]
Change-Id: Ic2748bdb55aa577da42243d525776b18eb9cd62a
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
[etienne: cherry-picked from f164f0f834203bb3d85f7f07eb2a2537d401a179]
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 core/drivers/regulator/regulator_gpio.c | 261 ++++++++++++++++++++++++
 core/drivers/regulator/sub.mk           |   1 +
 mk/config.mk                            |   6 +
 3 files changed, 268 insertions(+)
 create mode 100644 core/drivers/regulator/regulator_gpio.c

diff --git a/core/drivers/regulator/regulator_gpio.c b/core/drivers/regulator/regulator_gpio.c
new file mode 100644
index 000000000..c6fed63fb
--- /dev/null
+++ b/core/drivers/regulator/regulator_gpio.c
@@ -0,0 +1,261 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2023, STMicroelectronics
+ */
+
+#include <assert.h>
+#include <compiler.h>
+#include <drivers/gpio.h>
+#include <drivers/regulator.h>
+#include <dt-bindings/gpio/gpio.h>
+#include <kernel/delay.h>
+#include <libfdt.h>
+#include <trace.h>
+
+static_assert(GPIO_LEVEL_HIGH == 1 && GPIO_LEVEL_LOW == 0);
+
+/*
+ * struct regulator_gpio - GPIO controlled regulator
+ * @regulator: Preallocated regulator instance
+ * @enable_gpio: GPIO for the enable state of the regulator or NULL if always on
+ * @enable_delay: Time (in microsecond) for the regulator to get enabled
+ * @voltage_gpio: GPIO for the voltage level selection
+ * @voltage_levels_uv: 2 cells array supported voltage levels, increasing order
+ * @voltage_level_high: True if higher voltage level relates to active high GPIO
+ */
+struct regulator_gpio {
+	struct regulator regulator;
+	struct gpio *enable_gpio;
+	unsigned int enable_delay;
+	struct gpio *voltage_gpio;
+	int voltage_levels_uv[2];
+	bool voltage_level_high;
+};
+
+static struct regulator_gpio *regulator_priv(struct regulator *regulator)
+{
+	return container_of(regulator, struct regulator_gpio, regulator);
+}
+
+static TEE_Result regulator_gpio_set_state(struct regulator *regulator,
+					   bool enabled)
+{
+	struct regulator_gpio *regu = regulator_priv(regulator);
+
+	if (regu->enable_gpio) {
+		if (enabled) {
+			gpio_set_value(regu->enable_gpio, GPIO_LEVEL_HIGH);
+			if (regu->enable_delay)
+				udelay(regu->enable_delay);
+		} else {
+			gpio_set_value(regu->enable_gpio, GPIO_LEVEL_LOW);
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result regulator_gpio_read_state(struct regulator *regulator,
+					    bool *enabled)
+{
+	struct regulator_gpio *regu = regulator_priv(regulator);
+
+	if (regu->enable_gpio)
+		*enabled = gpio_get_value(regu->enable_gpio);
+	else
+		*enabled = true;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result regulator_gpio_set_voltage(struct regulator *regulator,
+					     int level_uv)
+{
+	struct regulator_gpio *regu = regulator_priv(regulator);
+
+	if (level_uv == regu->voltage_levels_uv[0])
+		gpio_set_value(regu->voltage_gpio, GPIO_LEVEL_LOW);
+	else if (level_uv == regu->voltage_levels_uv[1])
+		gpio_set_value(regu->voltage_gpio, GPIO_LEVEL_HIGH);
+	else
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result regulator_gpio_read_voltage(struct regulator *regulator,
+					      int *level_uv)
+{
+	struct regulator_gpio *regu = regulator_priv(regulator);
+
+	*level_uv = regu->voltage_levels_uv[gpio_get_value(regu->voltage_gpio)];
+
+	return TEE_SUCCESS;
+}
+
+static const struct regulator_ops regulator_gpio_ops = {
+	.set_state = regulator_gpio_set_state,
+	.get_state = regulator_gpio_read_state,
+	.set_voltage = regulator_gpio_set_voltage,
+	.get_voltage = regulator_gpio_read_voltage,
+};
+
+static TEE_Result get_enable_gpio(const void *fdt, int node,
+				  struct regulator_gpio *regu)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	const fdt32_t *cuint = NULL;
+	struct gpio *gpio = NULL;
+
+	res = gpio_dt_get_by_index(fdt, node, 0, "enable", &gpio);
+	if (res == TEE_ERROR_ITEM_NOT_FOUND) {
+		regu->enable_gpio = NULL;
+
+		return TEE_SUCCESS;
+	}
+	if (res)
+		return res;
+
+	/* Override active level phandle flag, as per DT bindings */
+	if (dt_have_prop(fdt, node, "enable-active-high"))
+		gpio->dt_flags &= ~GPIO_ACTIVE_LOW;
+	else
+		gpio->dt_flags |= GPIO_ACTIVE_LOW;
+
+	/* Override open drain/open source phandle flag, as per DT bindings */
+	if (dt_have_prop(fdt, node, "gpio-open-drain"))
+		gpio->dt_flags |= GPIO_LINE_OPEN_DRAIN;
+	else
+		gpio->dt_flags &= ~GPIO_LINE_OPEN_DRAIN;
+
+	cuint = fdt_getprop(fdt, node, "startup-delay-us", NULL);
+	if (cuint)
+		regu->enable_delay = fdt32_to_cpu(*cuint);
+
+	gpio_set_direction(gpio, GPIO_DIR_OUT);
+
+	regu->enable_gpio = gpio;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_voltage_level_gpio(const void *fdt, int node,
+					 struct regulator_gpio *regu)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	const fdt32_t *cuint = NULL;
+	struct gpio *gpio = NULL;
+	void *gpio_ref = &gpio;
+	int len = 0;
+
+	/*
+	 * DT bindings allows more than 1 GPIO to control more than
+	 * 2 voltage levels. As it's not used so far in known platforms
+	 * this implementation is simplified to support only 2 voltage
+	 * levels controlled with a single GPIO.
+	 */
+	if (dt_driver_device_from_node_idx_prop("gpios", fdt, node, 1,
+						DT_DRIVER_GPIO, gpio_ref) !=
+	    TEE_ERROR_ITEM_NOT_FOUND) {
+		EMSG("Multiple GPIOs not supported for level control");
+		return TEE_ERROR_GENERIC;
+	}
+
+	res = dt_driver_device_from_node_idx_prop("gpios", fdt, node, 0,
+						  DT_DRIVER_GPIO, gpio_ref);
+	if (res)
+		return res;
+
+	cuint = fdt_getprop(fdt, node, "states", &len);
+	if (!cuint || len != 4 * sizeof(fdt32_t)) {
+		EMSG("Node %s expects 2 levels from property \"states\"",
+		     fdt_get_name(fdt, node, NULL));
+		return TEE_ERROR_GENERIC;
+	}
+
+	if (fdt32_to_cpu(*(cuint + 1))) {
+		assert(!fdt32_to_cpu(*(cuint + 3)));
+		regu->voltage_levels_uv[0] = fdt32_to_cpu(*(cuint + 2));
+		regu->voltage_levels_uv[1] = fdt32_to_cpu(*(cuint));
+	} else {
+		assert(fdt32_to_cpu(*(cuint + 3)) == 1);
+		regu->voltage_levels_uv[0] = fdt32_to_cpu(*(cuint));
+		regu->voltage_levels_uv[1] = fdt32_to_cpu(*(cuint + 2));
+		regu->voltage_level_high = true;
+	}
+
+	gpio_set_direction(gpio, GPIO_DIR_OUT);
+
+	regu->voltage_gpio = gpio;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result regulator_gpio_probe(const void *fdt, int node,
+				       const void *compat_data __unused)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct regulator_gpio *regu = NULL;
+	struct regu_dt_desc desc = { };
+	const char *supply_name = NULL;
+	const char *type = NULL;
+	char *regu_name = NULL;
+
+	regu_name = (char *)fdt_get_name(fdt, node, NULL);
+
+	type = fdt_getprop(fdt, node, "regulator-type", NULL);
+	if (type && strcmp(type, "voltage")) {
+		EMSG("Regulator node %s: type \"%s\" not supported",
+		     regu_name, type);
+		res = TEE_ERROR_GENERIC;
+		goto err;
+	}
+
+	regu = calloc(1, sizeof(*regu));
+	if (!regu) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto err;
+	}
+
+	res = get_enable_gpio(fdt, node, regu);
+	if (res)
+		goto err;
+
+	res = get_voltage_level_gpio(fdt, node, regu);
+	if (res)
+		goto err;
+
+	if (fdt_getprop(fdt, node, "vin-supply", NULL))
+		supply_name = "vin";
+
+	desc = (struct regu_dt_desc){
+		.name = regu_name,
+		.ops = &regulator_gpio_ops,
+		.supply_name = supply_name,
+		.regulator = &regu->regulator,
+	};
+
+	res = regulator_dt_register(fdt, node, node, &desc);
+	if (res) {
+		EMSG("Can't register regulator %s: %#"PRIx32, regu_name, res);
+		goto err;
+	}
+
+	return TEE_SUCCESS;
+
+err:
+	free(regu);
+
+	return res;
+}
+
+static const struct dt_device_match regulator_gpio_match_table[] = {
+	{ .compatible = "regulator-gpio" },
+	{ }
+};
+
+DEFINE_DT_DRIVER(regulator_gpio_dt_driver) = {
+	.name = "regulator-gpio",
+	.match_table = regulator_gpio_match_table,
+	.probe = regulator_gpio_probe,
+};
diff --git a/core/drivers/regulator/sub.mk b/core/drivers/regulator/sub.mk
index 1562241c7..a00dd0192 100644
--- a/core/drivers/regulator/sub.mk
+++ b/core/drivers/regulator/sub.mk
@@ -1,4 +1,5 @@
 srcs-y += regulator.c
 srcs-$(CFG_DT) += regulator_dt.c
 srcs-$(CFG_REGULATOR_FIXED) += regulator_fixed.c
+srcs-$(CFG_REGULATOR_GPIO) += regulator_gpio.c
 srcs-$(CFG_STM32_VREFBUF) += stm32_vrefbuf.c
diff --git a/mk/config.mk b/mk/config.mk
index 820d06ea2..8c4b7e4f2 100644
--- a/mk/config.mk
+++ b/mk/config.mk
@@ -889,11 +889,17 @@ CFG_DRIVERS_PINCTRL ?= n
 #
 # When enabled, CFG_REGULATOR_FIXED embeds a voltage regulator driver for
 # DT compatible "regulator-fixed" devices.
+#
+# When enabled, CFG_REGULATOR_GPIO embeds a voltage regulator driver for
+# DT compatible "regulator-gpio" devices.
 CFG_DRIVERS_REGULATOR ?= n
 CFG_REGULATOR_FIXED ?= n
+CFG_REGULATOR_GPIO ?= n
 
 $(eval $(call cfg-enable-all-depends,CFG_REGULATOR_FIXED, \
 	 CFG_DRIVERS_REGULATOR CFG_DT))
+$(eval $(call cfg-enable-all-depends,CFG_REGULATOR_GPIO, \
+	 CFG_DRIVERS_REGULATOR CFG_DT CFG_DRIVERS_GPIO))
 
 # When enabled, CFG_INSECURE permits insecure configuration of OP-TEE core
 # and shows a print (info level) when booting up the device that
-- 
2.47.2

